ARROW-14753: [Doc] Steps in making your first PR - building C++ (#11820)
apache/arrow: docs/source/developers/guide/step_by_step/building.rst

.. _build-arrow:

************************************
Building the Arrow's libraries 🏋🏿‍♀️
************************************

The Arrow project contains good number of libraries that enable
work in many languages. Most libraries (C++, C#, Go, Java,
JavaScript, Julia, and Rust) already contain distinct implementations
of Arrow. 

It is different for C (Glib), MATLAB, Python, R, and Ruby as they
are built on top of the C++ library. In this section we will be

> dealing with this second part of the libraries plus some of C++.

Reviewer: ```suggestion
dealing with some of these libraries as well has how they work with the C++ library.
```

> In this case, if you decide to contribute to Arrow you will meet
> the topic of compiling the source code and use of CMake. You may
> have some experience with it or not. If not, it is good to read
> through this part so you understand what is happening in the process
> of building Arrow better.

Reviewer: ```suggestion
If you decide to contribute to Arrow you might
need to compile the C++ source code.  This is done using a tool called CMake, which you may or may not have experience with.  If not, this section of the guide will help you better understand CMake and the process
of building Arrow's C++ code.
```

Not all contributors will need to compile the C++; in the R developer docs we have instructions for using the nightly builds as the C++ source when people are only making changes to the R code without needing to build C++ from source.

If you feel comfortable with compiling then feel free to proceed
to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.

Building Arrow C++
==================

Why build Arrow C++ from source?
--------------------------------

> For Arrow C++ implementation and its bindings (Python and R for example)
> these bindings are wrapping the underlying C++ functions. Even if you
> want to work on PyArrow or R package the source code of C++ may have to
> be edited also.

Reviewer: ```suggestion
For Arrow implementations which are built on top of the C++ implementation
(e.g. Python and R), wrappers and interfaces have been written to the 
underlying C++ functions. If you want to work on PyArrow or R package, 
you may need to edit the source code of the C++ library too.
```

About CMake
-----------

CMake is a cross platform build system generator and it defers
to another program such as ``make`` or ``ninja`` for the actual build.
If running into errors with the build process, the first thing to try is
to tweak some CMake flags for compiling Arrow.

.. TODO
.. CMake presets
.. Ex: "We now have CMake Presets which are useful starting points."
.. https://github.com/apache/arrow/blob/master/docs/source/developers/cpp/building.rst#cmake-presets

Optional flags and why might we use them
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. TODO short description of the use of flags
.. R and Python have specific lists of flags in their respective building docs that can be referenced.

.. seealso::
	Full list of optional flags: :ref:`cpp_build_optional_components`

.. Environment variables useful for developers
.. ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. TODO short description of the use of env vars

Building from source vs. using binaries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Using binaries is a fast and simple way of working with the latest

Author: Thanks for pointing it out! I will rephrase.

Arrow version. But you do not have the possibility to add or change
the code and as a contributor you will need to.

Detailed instructions on building C++ library from source can
be found :ref:`here <building-arrow-cpp>`.

.. _build-pyarrow:

Building PyArrow
================

After building Arrow C++ part of Arrow you have to build PyArrow on top

> of it also. The reason is the same, so you can edit the code and run

Reviewer: ```suggestion
of it also. The reason is the same; so you can edit the code and run
```

tests on the edited code you have locally.

**Why do we have to do builds separately?**

> As mentioned in the beginning of this page, Python part of the Arrow

Reviewer: ```suggestion
As mentioned at the beginning of this page, the Python part of the Arrow
```

> project is built on top of C++. In order to make changes in Python part

Reviewer: ```suggestion
project is built on top of the C++ library. In order to make changes in the Python part
```

> of Arrow as well as C++ part of Arrow, we need to build them separately.

Reviewer: ```suggestion
of Arrow as well as the C++ part of Arrow, you need to build them separately.
```

We hope this introduction was enough to help you start with the building
process.

.. seealso::
	Follow the instructions to build PyArrow together with the C++ library

	- :ref:`build_pyarrow`
	Or

	- :ref:`build_pyarrow_win`

.. _build-rapackage:

> Building R package

Reviewer: ```suggestion
Building the R package
```

.. _build-arrow:

**********************************
Building the Arrow libraries 🏋🏿‍♀️
**********************************

The Arrow project contains a number of libraries that enable
work in many languages. Most libraries (C++, C#, Go, Java,
JavaScript, Julia, and Rust) already contain distinct implementations
of Arrow. 

It is different for C (Glib), MATLAB, Python, R, and Ruby as they
are built on top of the C++ library. In this section of the guide

> we will try to make a friendly introduction to the build

Reviewer: ```suggestion
we will try to make a friendly introduction to the build,
```

dealing with some of these libraries as well has how they work with
the C++ library.

If you decide to contribute to Arrow you might need to compile the
C++ source code. This is done using a tool called CMake, which you
may or may not have experience with. If not, this section of the
guide will help you better understand CMake and the process
of building Arrow's C++ code.

> If you feel comfortable with compiling then feel free to proceed
> to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
> `R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.

Reviewer: ```suggestion
This content is intended to help explain the concepts related to 
and tools required for building Arrow's C++ library from source.
If you are looking for the specific required steps, or already feel comfortable 
with compiling Arrow's C++ library, then feel free to proceed
to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.
```
How about we add a little bit here to emphasise that this section is about *understanding* and isn't intended to provide concrete steps for the reader?

Building Arrow C++
==================

Why build Arrow C++ from source?
--------------------------------

For Arrow implementations which are built on top of the C++ implementation
(e.g. Python and R), wrappers and interfaces have been written to the

> underlying C++ functions. If you want to work on PyArrow or R package,

Reviewer: ```suggestion
underlying C++ functions. If you want to work on PyArrow or the R package,
```

starting points.

More detailed information about CMake presets can be found in
the :ref:`cmake_presets` section.

Optional flags and environment variables
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Flags used in the CMake build are used to include additional components
and to handle third-party dependencies.
The build for C++ library can be minimal with no use of flags or can
be changed with adding optional components from the
:ref:`list <cpp_build_optional_components>`.

.. seealso::
	Full list of optional flags: :ref:`cpp_build_optional_components`

R and Python have specific lists of flags in their respective builds
that need to be included. You can find the links at the end
of this section.

In general on Python side the options are set with CMake flags and
paths with environment variables. In R the environment variables are used
for all things connected to the build, also for setting CMake flags.

Building from source vs. using binaries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Using binaries is a fast and simple way of working with the last release
of Arrow. However, if you use these it means that you will be unable to
make changes to the Arrow C++ library.

> **Note:** every language has it's own way of dealing with binaries.

Reviewer: ```suggestion
**Note:** every language has its own way of dealing with binaries.
```

.. _build-arrow:

**********************************
Building the Arrow libraries 🏋🏿‍♀️
**********************************

The Arrow project contains a number of libraries that enable
work in many languages. Most libraries (C++, C#, Go, Java,
JavaScript, Julia, and Rust) already contain distinct implementations
of Arrow. 

> It is different for C (Glib), MATLAB, Python, R, and Ruby as they

Reviewer 2: ```suggestion
This is different for C (Glib), MATLAB, Python, R, and Ruby as they
```

of building Arrow's C++ code.

If you feel comfortable with compiling then feel free to proceed
to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.

Building Arrow C++
==================

Why build Arrow C++ from source?
--------------------------------

For Arrow implementations which are built on top of the C++ implementation
(e.g. Python and R), wrappers and interfaces have been written to the
underlying C++ functions. If you want to work on PyArrow or R package,
you may need to edit the source code of the C++ library too.

Detailed instructions on building C++ library from source can
be found :ref:`here <building-arrow-cpp>`.

About CMake
-----------

CMake is a cross-platform build system generator and it defers
to another program such as ``make`` or ``ninja`` for the actual build.
If you are running into errors with the build process, the first thing to
do is to look at the error message thoroughly and check the building
documentation for any similar error advice. Also changing the CMake flags
for compiling Arrow could be useful.

CMake presets

> -------------

Reviewer 2: ```suggestion
^^^^^^^^^^^^^^^^^^^
```

(then it's a subsection below "CMake"?)

You could also try to build with CMake presets which are a collection of
build and test recipes for Arrow's CMake. They are a very useful
starting points.

More detailed information about CMake presets can be found in
the :ref:`cmake_presets` section.

Optional flags and environment variables
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Flags used in the CMake build are used to include additional components
and to handle third-party dependencies.
The build for C++ library can be minimal with no use of flags or can
be changed with adding optional components from the
:ref:`list <cpp_build_optional_components>`.

.. seealso::
	Full list of optional flags: :ref:`cpp_build_optional_components`

R and Python have specific lists of flags in their respective builds
that need to be included. You can find the links at the end
of this section.

> In general on Python side the options are set with CMake flags and

Reviewer 2: ```suggestion
In general on Python side, the options are set with CMake flags and
```

> Building Arrow's libraries 🏋🏿‍♀️

Reviewer 3: It's not in this PR, but this should probably be "Building the Arrow libraries".

*********************************

If you decide to contribute to Arrow you will meet the topic of

> compiling source code and use of CMake. You may have some

Reviewer 3: The mention of CMake applies mostly to Arrow C++ and C++-derived bindings such as Python or R (not Java or Go, for example). Perhaps you could start this document by mentioning this context?

Author: Oh, didn't even know that. Will have to learn about other libraries to be able to explain the difference. Thanks for pointing it out!

Author: Well, it's nicely explained in https://arrow.apache.org/overview/ 🤦‍♀️

> compiling source code and use of CMake. You may have some

Reviewer 2: ```suggestion
compiling the source code and use of CMake. You may have some
```

experience with it or not. If not, it is good to read through
this part so you understand what is happening in the process of
building Arrow better.

If you feel comfortable with compiling then feel free to proceed
to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.

Building C++
============

> Why build C++ from source?

Reviewer 3: "Arrow C++", not just "C++" (and in the title above as well) :-)

--------------------------

> The core of Arrow is written in C++ and all work done in other

Reviewer 4: ```suggestion
The core of Arrow is written in C++ and all bindings in other
```

> languages (Python, R, ..) is wrapping-binding-connecting underlying

Reviewer 4: ```suggestion
languages (Python, R, ..) are wrapping underlying
```

> C++ functions. Even if you want to work on PyArrow or R package

Reviewer 2: This is only true for the C++ implementation and its bindings. We should maybe try to reword this a bit to more correctly represent the fact that there are non-C++ Arrow implementations

the source code of C++ may have to be edited also.

About CMake
-----------

> CMake is a cross platform build system generator and it uses make

Reviewer 2: The "uses make for the actual build" actually depends on the cmake settings. For example we recommend to use ninja instead of make.

> for the actual build. In the compiling process of Arrow what will

Reviewer 3: Hmm, it doesn't necessarily use make (we actually recommend [Ninja](https://ninja-build.org/) in the C++ build docs).

Therefore you could rephrase this as : "... and it defers to another program such as `make` or `ninja` for the actual build".

> for the actual build. In the compiling process of Arrow what will

Reviewer 2: ```suggestion
for the actual build. In the compilation process of Arrow what will
```

most probably be needed is some tweaking of the flags that are added

> to cmake in the compiling process of Arrow.

Reviewer 3: "CMake"

> to cmake in the compiling process of Arrow.

Reviewer 2: This sentence is a bit hard to read. Maybe something like "You will probably need to tweak some CMake flags for compiling Arrow."

Optional flags and why might we use them
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> .. TODO short description of the use of flags

Reviewer 5: Two things worth mentioning here:

 1. R and Python have specific lists of flags in their respective building docs that can be referenced.
 2. We now have [CMake Presets](https://github.com/apache/arrow/blob/master/docs/source/developers/cpp/building.rst#cmake-presets) which are useful starting points.

.. seealso::
	Full list of optional flags: :ref:`cpp_build_optional_components`

.. Environment variables useful for developers
.. ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. TODO short description of the use of env vars

Building from source vs. using binaries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Using binaries is a fast and simple way of working with the latest
Arrow version. But you do not have the possibility to add or change
the code and as a contributor you will need to.

Detailed instructions on building C++ library from source can
be found :ref:`here <building-arrow-cpp>`.

.. _build-pyarrow:

Building PyArrow
================

> After building C++ part of Arrow you have to build PyArrow on top of it

Reviewer 3: "the Arrow C++ libraries"

> After building C++ part of Arrow you have to build PyArrow on top of it

Reviewer 2: ```suggestion
After building the C++ part of Arrow you have to build PyArrow on top of it
```

also. The reason is the same, so you can edit the code and run tests on
the edited code you have locally.

> **Why do we have to do builds separately?**

Reviewer 3: I'm not sure it's worth bothering beginners with this question :-) If we want to, though, we'll need to find a better wording for the answer, as currently it feels a bit confusing.

Author: Yeah, showing I am still a little confused :P Will try to rephrase - as I would love to have this info when I started ;)
